[jax2tf] Initial support for kwargs.

saeta opened the pull request:
Fixes #6791

saeta committed:
[jax2tf] Initial support for kwargs.

Fixes #6791

saeta commented:
@gnecula PTAL? I'm sure there's lots I'm doing wrong; feedback welcome! :-)

gnecula commented on jax/experimental/jax2tf/jax2tf.py:
```diff
@@ -206,19 +206,23 @@ def check_arg(a):
     if polymorphic_shapes is None:
       polymorphic_shapes_ = (None,) * len(args)
     else:
+      if kwargs:
```
I think that what may be better to do here is to say that we support polymorphic_shapes only for the positional arguments, and do not raise this error. The check in line 226 should still work, and your changes below in line 236 should take care of setting the kwargs to be shape monomorphic. So, the only thing to do is to drop this check here, and update the docstring.

BTW, this suggestion follows the model of what we do for vmap, where we have an `in_axes` (they apply only to positional arguments).

saeta replied:
Excellent. That's great.

gnecula commented on jax/experimental/jax2tf/tests/jax2tf_test.py:
```diff
@@ -516,6 +516,15 @@ def jax_fn_array(x):
 
+  def test_kwargs(self):
+    # Re: https://github.com/google/jax/issues/6791
+    def f_jax(x):
```
Would this test fail if we replace `x` with `y`? Probably not. Maybe the test would be clearer if `x` is forced to be a kwarg here?

saeta replied:
I've now forced `x` to be a `kwarg` by adding a `*`.

saeta committed:
Respond to comments.

saeta commented:
PTAL? In particular, feedback regarding the text in the docstring would be appreciated!

gnecula committed:
Update jax2tf.py

Minor typo fix

gnecula committed:
Update jax2tf.py

gnecula reviewed (approved):
Thank you!

gnecula commented on jax/experimental/jax2tf/tests/jax2tf_test.py:
```diff
@@ -516,6 +516,15 @@ def jax_fn_array(x):
+      return jnp.sum(x)
+    f_tf = jax2tf.convert(f_jax)
+    self.assertAllClose(
+      f_tf(x=jnp.zeros(3)),  # Call with kwargs.
```
This test fails with JAX_ENABLE_X64=True. I think that you may need to force `jnp.zeros(3, dtype=jnp.float32)`. 

Note that unfortunately the jax2tf test do not run in GitHub, only internally, due to JAX not being ready to take a dependency on tf_nightly, which is needed for jax2tf. This means that test failures only pop up when we try to import the change in Google.

gnecula committed:
Update jax2tf_test.py

Fix the type of np.zeros to be float32. 
Also, replaced `jnp.zeros` with `np.zeros` because technically the argument to `f_tf` should be a TF value, not a JAX value.

gnecula committed:
Merge branch 'master' into bug-6791

gnecula committed:
Update jax2tf_test.py
